pkg/controller/atlasproject/project_settings.go

```diff
@@ -0,0 +1,95 @@
+package atlasproject
+
+import (
+	"context"
+	"reflect"
+
+	v1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+)
+
+func ensureProjectSettings(ctx *workflow.Context, projectID string, project *v1.AtlasProject) (result workflow.Result) {
+	if result = syncProjectSettings(ctx, projectID, project); !result.IsOk() {
+		ctx.SetConditionFromResult(status.ProjectSettingsReadyType, result)
+		return result
+	}
+
+	if project.Spec.Settings == nil {
+		ctx.UnsetCondition(status.ProjectSettingsReadyType)
+		return workflow.OK()
+	}
+
+	ctx.SetConditionTrue(status.ProjectSettingsReadyType)
+	return workflow.OK()
+}
+
+func syncProjectSettings(ctx *workflow.Context, projectID string, project *v1.AtlasProject) workflow.Result {
+	spec := project.Spec.Settings
+
+	atlas, err := fetchSettings(ctx, projectID)
+	if err != nil {
+		return workflow.Terminate(workflow.ProjectSettingsReady, err.Error())
+	}
+
+	if !areSettingsInSync(atlas, spec) {
+		if err := patchSettings(ctx, projectID, spec); err != nil {
+			return workflow.Terminate(workflow.ProjectSettingsReady, err.Error())
+		}
+	}
+
+	return workflow.OK()
+}
+
+func areSettingsInSync(atlas, spec *v1.ProjectSettings) bool {
+	return isOneContainedInOther(spec, atlas)
+}
+
+func patchSettings(ctx *workflow.Context, projectID string, spec *v1.ProjectSettings) error {
+	specAsAtlas, err := spec.ToAtlas()
+	if err != nil {
+		return err
+	}
+
+	_, _, err = ctx.Client.Projects.UpdateProjectSettings(context.Background(), projectID, specAsAtlas)
+	return err
+}
+
+func fetchSettings(ctx *workflow.Context, projectID string) (*v1.ProjectSettings, error) {
+	data, _, err := ctx.Client.Projects.GetProjectSettings(context.Background(), projectID)
+	if err != nil {
+		return nil, err
+	}
+	ctx.Log.Debugw("Got Project Settings", "data", data)
+
+	settings := v1.ProjectSettings(*data)
+	return &settings, nil
+}
+
+func isOneContainedInOther(one, other *v1.ProjectSettings) bool {
```

Comment:
I don't fully understand the advantage of using the reflect package. This piece of code can be rewritten without it and more explicitly. For example:
```
func isOneContainedInOther(one, other *v1.ProjectSettings) bool {
	if one == nil {
		return true
	}

	if other == nil {
		return false
	}

	return compareBoolPtr(one.IsCollectDatabaseSpecificsStatisticsEnabled, other.IsCollectDatabaseSpecificsStatisticsEnabled) &&
		compareBoolPtr(one.IsDataExplorerEnabled, other.IsDataExplorerEnabled) &&
		compareBoolPtr(one.IsPerformanceAdvisorEnabled, other.IsPerformanceAdvisorEnabled) &&
		compareBoolPtr(one.IsRealtimePerformancePanelEnabled, other.IsRealtimePerformancePanelEnabled)
}

func compareBoolPtr(oneBool, otherBool *bool) bool {
	if oneBool == nil {
		return true
	}
	if otherBool == nil {
		return false
	}
	return *oneBool == *otherBool
}
```

Reply:
Imagine if they add 10 more fields to the Project Settings 🤔 
The "beauty" of your solution is dependent on the amount of fields in the struct so I think it's not as good :)

Reply:
Also at some point in the future isOneContainedInOther can be moved out like this:
```
func isOneContainedInOther[T](one, other *T) bool
```
And it can potentially be reused for other stuff :)

Reply:
But there are four fields in this structure. And if it changes, maybe there will be not only bool ptrs. 
And I don't think it's a good idea to use this approach for any type. Because this would require a lot of complicated code with type checks and other reflect magic.

Reply:
Well, there is some logic in what you're saying)

In general I think it makes sense to create one general function (even with complex logic inside) if it allows not to call simple function on every single struct field :)

Reply:
At the moment, this function is named very generically. At the same time, this is a tricky comparator that works with only one type.
Also, I do not understand why you use the reflection package here. Because it is currently needed to save 3 lines of code, but at the same time it brings several methods that replace well-known operators. And it does not help to reduce the code, nor to make this code more readable.
And I do not know of other cases in which this function would be useful in its current form.

Reply:
The function logic is actually quite simple. We need to iterate over struct fields and `reflect` is the only way I know of in `Go`.

At this point it is an unexported function with types hardcoded so there is no expectation that it will be reused in its current form. I see no reason to over-engineer this function now so that we can reuse it later. I am just saying that since the taken `reflect-approach` is quite a general one, similar logic can be reused in the future in the exported utils func.